File: services/api/kinetica/auth/canonical.py
"""Canonical JSON serialization used for `bootstrap_signature` (and
later flows). Must produce byte-identical output to the TS client at
`apps/mobile/src/auth/canonical.ts` for any given dict — those bytes
are the exact input the device's Ed25519 key signs over (after
SHA-256), so any disagreement breaks the signature.

Format:
  - sorted keys
  - separators (',', ':') — no whitespace
  - UTF-8 encoded
  - non-ASCII characters emitted directly as their UTF-8 bytes
    (NOT escaped to `\\uXXXX`). JS's `JSON.stringify` does this by
    default; Python's `json.dumps` defaults to `ensure_ascii=True`
    which would diverge — `ensure_ascii=False` here matches the JS
    behavior. Fixture #4 in `docs/fixtures/canonical_json.json` is
    the canary for this case.

Byte fields (pubkeys, salts, ciphertext) are pre-encoded as standard
base64 strings before they reach this function, so the canonical form
is just a flat JSON object over strings.
"""

from __future__ import annotations

import hashlib
import json


def canonical_json(obj: dict[str, str]) -> bytes:
    return json.dumps(
        obj,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    ).encode("utf-8")


def canonical_sha256(obj: dict[str, str]) -> bytes:
    return hashlib.sha256(canonical_json(obj)).digest()

File: apps/mobile/src/auth/canonical.ts
// Canonical JSON serialization + SHA-256 used to build the input to
// `bootstrap_signature` (and to other request signatures introduced by
// later flows). Both client and server MUST produce byte-identical
// output for a given object — the signing input is whatever bytes this
// function emits, full stop.
//
// Format: sorted-key JSON with `(',', ':')` separators, UTF-8 encoded.
// All byte fields are pre-encoded as base64 strings before they reach
// here, so the canonical form is just standard JSON over a flat string
// dict. Server-side, `json.dumps(obj, sort_keys=True, separators=(",", ":"))`
// produces the matching output (see `services/api/kinetica/auth/canonical.py`).

import { sodium } from '../crypto/index.ts';

export function canonicalJson(obj: Record<string, string>): string {
  const sortedKeys = Object.keys(obj).sort();
  const pairs = sortedKeys.map((k) => `${JSON.stringify(k)}:${JSON.stringify(obj[k])}`);
  return `{${pairs.join(',')}}`;
}

export function sha256(message: Uint8Array): Uint8Array {
  return sodium().crypto_hash_sha256(message);
}
